senders/test_server_test.go

```diff
@@ -0,0 +1,52 @@
+package senders
+
+import (
+	"bufio"
+	"compress/gzip"
+	"net/http"
+	"net/http/httptest"
+)
+
+func startTestServer() *testServer {
+	handler := &testServer{}
+	server := httptest.NewServer(handler)
+	handler.httpServer = server
+	handler.URL = server.URL
+	return handler
+
+}
+
+type testServer struct {
+	MetricLines []string
+	httpServer  *httptest.Server
+	URL         string
+}
+
+func (s *testServer) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	newLines, err := decodeMetricLines(request)
+	if err != nil {
+		writer.WriteHeader(500)
+	}
+	for _, metricLine := range newLines {
+		s.MetricLines = append(s.MetricLines, metricLine)
+	}
+	writer.WriteHeader(200)
+}
+
+func decodeMetricLines(request *http.Request) ([]string, error) {
+	metricLines := make([]string, 0)
+	reader, err := gzip.NewReader(request.Body)
+	if err != nil {
+		return metricLines, err
+	}
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
```

Comment:
scanner can return errors through scanner.Err().  Have a look at https://pkg.go.dev/bufio#example-Scanner-Custom to see how.